### PR TITLE
Simplify ERC20

### LINKED ISSUE #109: Simplify ERC20 extension
### Discussed in https://github.com/OpenZeppelin/cairo-contracts/discussions/103

<div type='discussions-op-text'>

<sup>Originally posted by **martriay** December 14, 2021</sup>
I was thinking that maybe we could abstract ERC20 even further and move `transferFrom`, `increaseAllowance`, etc. implementations to the base contract, and require frontend contracts just to expose them instead of re-implementing them. What do you think?</div>

### PR DESCRIPTION
**Pull request by andrew-fleming**
# Simplify ERC20

## Summary

This PR includes a further abstraction of the extensibility pattern for ERC20 contracts and resolves #109 . This structure follows the concept that all of the necessary contract logic occurs within the base contract. Further, the frontend contracts simply import the requisite methods and expose them with the `view` and `external` decorators.

### CODE REVIEW

**Review by martriay — CHANGES_REQUESTED** [review]
Almost there, suggestions below :)

File: contracts/token/ERC20_base.cairo [hunk 1]
```diff
@@ -56,11 +57,10 @@ func ERC20_initializer{
 end
 
 #
-# Getters
+# Importables
 #
 
-@view
-func name{
+func ERC20_name_{
```
[the remark this answers is not in the document]
Comment (andrew-fleming) [reply contracts/token/ERC20_base.cairo:63]: Hmm I considered exporting the storage variables. I chose exportable setter methods for consistency, so I'm leaning that way—easy change if you think accessing the storage vars directly is the way to go here

File: contracts/token/ERC20_base.cairo [hunk 2]
```diff
@@ -119,70 +114,113 @@ func allowance{
...
-    # add to recipient
-    let (recipient_balance: Uint256) = ERC20_balances.read(account=recipient)
-    # overflow is not possible because sum is guaranteed by mint to be less than total supply
-    let (new_recipient_balance, _: Uint256) = uint256_add(recipient_balance, amount)
-    ERC20_balances.write(recipient, new_recipient_balance)
+    # validates new_allowance < current_allowance and returns 1 if true   
+    let (enough_allowance) = uint256_lt(new_allowance, current_allowance)
+    assert_not_zero(enough_allowance)
+
+    #ERC20_approve(caller, spender, new_allowance)
```
[the remark this answers is not in the document]
Comment (andrew-fleming) [reply contracts/token/ERC20_base.cairo:199]: Ahhh! It should not!

File: contracts/token/ERC20_base.cairo (hunk 1 above)
[the remark this answers is not in the document]
Comment (martriay) [reply contracts/token/ERC20_base.cairo]: I think it's an unneeded indirection, let's use storage vars.

**Review by martriay — APPROVED** [review]

**Comment by martriay** [comment]
Thanks!

**Pull request merged**